### ISSUE TITLE
Gen4AI Generator

### ISSUE THREAD

**Issue by dmriajul** [dmriajul/Gen4Pic#1]
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models
import os
import sys
import pandas as pd
import matplotlib.pyplot as plt
from flask import Flask, request, jsonify
from PIL import Image
import requests

# Configure GPU memory growth if available
physical_devices = tf.config.experimental.list_physical_devices('GPU')
if physical_devices:
    tf.config.experimental.set_memory_growth(physical_devices[0], True)

# Directory for generated images
OUTPUT_DIR = "gen4pic_output"
os.makedirs(OUTPUT_DIR, exist_ok=True)

# Log file for debugging
LOG_FILE = "gen4pic_debug.log"

def log_message(message):
    with open(LOG_FILE, "a") as log:
        log.write(f"{message}\n")

# ----------------------------
# Generator Model
# ----------------------------
def build_generator():
    model = models.Sequential(name="Generator")
    model.add(layers.Dense(4 * 4 * 512, activation='relu', input_shape=(100,)))
    model.add(layers.Reshape((4, 4, 512)))

    model.add(layers.Conv2DTranspose(256, kernel_size=5, strides=2, padding='same', activation='relu'))
    model.add(layers.Conv2DTranspose(128, kernel_size=5, strides=2, padding='same', activation='relu'))
    model.add(layers.Conv2DTranspose(64, kernel_size=5, strides=2, padding='same', activation='relu'))
    model.add(layers.Conv2DTranspose(3, kernel_size=5, strides=2, padding='same', activation='tanh'))
    
    return model

# ----------------------------
# Save Generated Images
# ----------------------------
def save_image(image_array, filename):
    try:
        image = (image_array * 127.5 + 127.5).astype(np.uint8)
        image = Image.fromarray(image)
        image.save(filename)
    except Exception as e:
        log_message(f"Error saving image {filename}: {e}")

# ----------------------------
# Flask API Integration
# ----------------------------
app = Flask(__name__)

generator = build_generator()
try:
    generator.load_weights("generator_weights.h5")  # Load pre-trained weights
except Exception as e:
    log_message(f"Error loading weights: {e}")
    sys.exit("Error: Cannot load generator weights.")

@app.route('/generate', methods=['POST'])
def generate():
    try:
        data = request.json
        prompt = data.get('prompt', 'default')  # Placeholder for custom prompts
        num_images = int(data.get('num_images', 1))

        log_message(f"Generating {num_images} images for prompt: {prompt}")

        # Generate images based on noise
        noise = np.random.normal(0, 1, (num_images, 100))
        generated_images = generator.predict(noise)

        image_urls = []
        for i, image in enumerate(generated_images):
            filename = os.path.join(OUTPUT_DIR, f"generated_{prompt}_{i}.png")
            save_image(image, filename)
            image_urls.append(f"https://github.com/khandokarriajulislam/blob/main/{filename}")

        log_message(f"Successfully generated images: {image_urls}")
        return jsonify({'image_urls': image_urls})

    except Exception as e:
        log_message(f"Error during generation: {e}")
        return jsonify({'error': str(e)})

if __name__ == '__main__':
    log_message("Starting Gen4Pic server...")
    try:
        app.run(debug=True, host='0.0.0.0', port=5000)
    except Exception as e:
        log_message(f"Server error: {e}")
        sys.exit("Error: Unable to start server.")
